Added Socrata API to retriever, Added tests for the socrata functions

Aakash3101 opened the pull request:
## Added the Socrata API
### Total number of datasets supported : 85,244 out of 213,965

- Added `-socrata` flag to `retriever ls` command. Use `retriever ls -socrata <name>` to display a list of suggestions of the dataset name you are looking for. Then, select the dataset name, and the details of the selected dataset name will be displayed.
Example : 
![Screenshot from 2021-06-30 23-31-45](https://user-images.githubusercontent.com/44918472/124009607-705b9180-d9fb-11eb-94da-e8c9669fcd10.png)

- Added `--socrata` flag to `retriever install` and `retriever download` commands. Use `retriever install --socrata <engine> <socrata-id>` to install the socrata dataset which is identified by that `socrata-id`, first it downloads the raw data, creates a script for the data, updates the script accordingly, and then install the script into the engine.
 And use `retriever download --socrata <engine> <socrata-id>` to download the raw data files of the socrata dataset identified by its `socrata-id`.
Example : 
- `retriever download`
![Screenshot from 2021-06-30 23-46-16](https://user-images.githubusercontent.com/44918472/124011264-75b9db80-d9fd-11eb-9579-91088cb38493.png)
- `retriever install`
![Screenshot from 2021-06-30 23-50-36](https://user-images.githubusercontent.com/44918472/124011818-0c869800-d9fe-11eb-8f79-03f6ec3c7598.png)

Added tests for all the Socrata functions.

## Notes:

- `retriever ls -socrata` with an empty input will display the `retriever ls -h -socrata` output. An input with a spelling error will return a list with 0 results.

- `retriever download --socrata <id>` with an incorrectly formatted id would display `Id <id> is incorrectly formatted and will not identify any asset.`

- `retriever download --socrata <id>`, with an `id` that identifies a unsupported dataset will display `Dataset cannot be downloaded using Socrata API because <datatype> type datasets are not supported`

- retriever currently supports only `tabular` type datasets, which excludes `{map: tabular}` datasets.

- Datasets that are `tabular` and do not have any column names in the metadata, will display the following for the `retriever install/download` command:
![Screenshot from 2021-07-01 00-09-46](https://user-images.githubusercontent.com/44918472/124014228-e0b8e180-da00-11eb-96e1-510167bd3421.png)

- If a Socrata dataset is already downloaded, it does not download the data again or its script exists, then the script creation process is not repeated. Example:
![Screenshot from 2021-07-01 00-14-04](https://user-images.githubusercontent.com/44918472/124015005-d5b28100-da01-11eb-9071-541a4a65ae88.png)

Aakash3101 commented:
`Service 'python_retriever' failed to build : The command '/bin/sh -c pip install psycopg2-binary -U' returned a non-zero code: 1` causes to fail the tests.

henrykironde commented on retriever/__main__.py:
```diff
@@ -122,6 +124,33 @@ def main():
+                    len(name_list)))
+                if len(name_list) > 0:
+                    question = [
+                        inquirer.List('dataset name',
```
Test select name on a smaller terminal 
retriever ls -socrata fishing

henrykironde replied:
Select and then cancel(control c)
Cancelled by user

```Python
Traceback (most recent call last):
  File "/Users/henrykironde/Documents/GitHub/python-virtual-environments/env/bin/retriever", line 11, in <module>
    load_entry_point('retriever==3.0.1.dev0', 'console_scripts', 'retriever')()
  File "/Users/henrykironde/Documents/GitHub/python-virtual-environments/env/lib/python3.7/site-packages/retriever/__main__.py", line 138, in main
    dataset_name = answer['dataset name']
TypeError: 'NoneType' object is not subscriptable
```

Aakash3101 commented:
## Updates
- Changed the `retriever ls -socrata` command to `retriever ls -s` command.

- Removed user prompt to name the script for the socrata dataset. The script name is of the format `socrata-<socrata_id>`. 
  E.g. `socrata-35s3-nmpm`.

- Shifted the `inquirer` module import in `defaults.py` into a try-except block. To run the `retriever ls -s` command, the user should install the `inquirer` module. An error message would pop up if `inquirer` is not installed while running the `retriever ls -s` command. Added the `KeyboardInterrupt` for `Ctrl+C` event while selecting names.

## Clarifications

- To see the help for `retriever install --socrata` command, run `retriever install --socrata <any-engine-name> -h`
```
$ retriever install --socrata postgres -h
usage: retriever install postgres [-h] [--hash-value HASH_VALUE]
                                  [-b BBOX BBOX BBOX BBOX] [--user [USER]]
                                  [--password [PASSWORD]] [--host [HOST]]
                                  [--port [PORT]] [--database [DATABASE]]
                                  [--database_name [DATABASE_NAME]]
                                  [--table_name [TABLE_NAME]]
                                  dataset

positional arguments:
  dataset               dataset name|socrata dataset identifer (xxxx-xxxx)

optional arguments:
  -h, --help            show this help message and exit
  --hash-value HASH_VALUE
                        install dataset from provenance directory using hash
                        value
  -b BBOX BBOX BBOX BBOX, --bbox BBOX BBOX BBOX BBOX
                        Set bounding box xmin, ymin, xmax, ymax
  --user [USER], -u [USER]
                        Enter your PostgreSQL username
  --password [PASSWORD], -p [PASSWORD]
                        Enter your password
  --host [HOST], -o [HOST]
                        Enter your PostgreSQL host
  --port [PORT], -r [PORT]
                        Enter your PostgreSQL port
  --database [DATABASE], -d [DATABASE]
                        Enter your PostgreSQL database name
  --database_name [DATABASE_NAME], -a [DATABASE_NAME]
                        Format of schema name
  --table_name [TABLE_NAME], -t [TABLE_NAME]
                        Format of table name
```
- Similarly for `retriever download --socrata` command, run `retriever download --socrata -h`.
```
$ retriever download --socrata -h                                                                                                                                                         
usage: retriever download [-h] [--debug] [--not-cached] [--socrata]
                          [-b BBOX BBOX BBOX BBOX] [--path [PATH]]
                          [--sub_dir [SUB_DIR]]
                          dataset

positional arguments:
  dataset               dataset name|socrata dataset identifer (xxxx-xxxx)

optional arguments:
  -h, --help            show this help message and exit
  --debug               run in debug mode
  --not-cached          overwrites local cache of raw data
  --socrata             installs datasets hosted on Socrata API
  -b BBOX BBOX BBOX BBOX, --bbox BBOX BBOX BBOX BBOX
                        Set bounding box xmin, ymin, xmax, ymax
  --path [PATH], -p [PATH]
                        File path to copy data files
  --sub_dir [SUB_DIR], -s [SUB_DIR]
                        Install directory
```
---
## Interface examples

1. `socrata_autocomplete_search()`
The input argument should be a list of strings. The return type is also a list of strings which are the autocompleted names.
```python
>>> import retriever as rt
>>> names = rt.socrata_autocomplete_search(['clinic', '2015', '2016'])
>>> for name in names:
...     print(name)
... 
2016 & 2015 Clinic Quality Comparisons for Clinics with Five or More Service Providers
2015 - 2016 Clinical Quality Comparison (>=5 Providers) by Geography
2016 & 2015 Clinic Quality Comparisons for Clinics with Fewer than Five Service Providers

```
2. `socrata_dataset_info()`
The input argument should be a string (valid dataset name returned by `socrata_autocomplete_search`). The function returns a list of dicts, because there are multiple datasets on socrata with same name (e.g. Building Permits).
```python
>>> import retriever as rt
>>> resource = rt.socrata_dataset_info('2016 & 2015 Clinic Quality Comparisons for Clinics with Five or More Service Providers')
>>> from pprint import pprint
>>> pprint(resource)
[{'description': 'This data set includes comparative information for clinics '
                 'with five or more physicians for medical claims in 2015 - '
                 '2016. \r\n'
                 '\r\n'
                 'This data set was calculated by the Utah Department of '
                 'Health, Office of Healthcare Statistics (OHCS) using Utah’s '
                 'All Payer Claims Database (APCD).',
  'domain': 'opendata.utah.gov',
  'id': '35s3-nmpm',
  'link': 'https://opendata.utah.gov/Health/2016-2015-Clinic-Quality-Comparisons-for-Clinics-w/35s3-nmpm',
  'name': '2016 & 2015 Clinic Quality Comparisons for Clinics with Five or '
          'More Service Providers',
  'type': {'dataset': 'tabular'}}]
```
3. `find_socrata_dataset_by_id()`
The input argument of the function should be the four-by-four socrata dataset identifier (e.g. `35s3-nmpm`). The function returns a dict which contains metadata about the dataset.
```python
>>> import retriever as rt
>>> from pprint import pprint
>>> resource = rt.find_socrata_dataset_by_id('35s3-nmpm')
>>> pprint(resource)
{'datatype': 'tabular',
 'description': 'This data set includes comparative information for clinics '
                'with five or more physicians for medical claims in 2015 - '
                '2016. \r\n'
                '\r\n'
                'This data set was calculated by the Utah Department of '
                'Health, Office of Healthcare Statistics (OHCS) using Utah’s '
                'All Payer Claims Database (APCD).',
 'domain': 'opendata.utah.gov',
 'homepage': 'https://opendata.utah.gov/Health/2016-2015-Clinic-Quality-Comparisons-for-Clinics-w/35s3-nmpm',
 'id': '35s3-nmpm',
 'keywords': ['socrata'],
 'name': '2016 & 2015 Clinic Quality Comparisons for Clinics with Five or More '
         'Service Providers'}
```
4. `update_socrata_contents(json_file, script_name, url, resource)`
This function updates the contents of the socrata script created by `install_socrata_dataset()`. The input arguments are:
- json_file = The content of the script created
- script_name =  The name of the script
- url = The url through which the dataset is downloaded
- resource = The object returned by the `find_socrata_dataset_by_id()`
The function returns `True, json_file` if the resource dict is correct, otherwise, it returns `False, None`.
```python
import retriever as rt
import json
from retriever.lib.defaults import SOCRATA_SCRIPT_WRITE_PATH

resource = rt.find_socrata_dataset_by_id('35s3-nmpm')
filename = resource["id"] + '.csv'
url = 'https://' + resource["domain"] + '/resource/' + filename
script_name = 'socrata-35s3-nmpm'
script_path = SOCRATA_SCRIPT_WRITE_PATH
script_filename = script_name.replace("-","_") + ".json"
with open(f"{script_path}/{script_filename}", "r") as f:
       json_file = json.load(f)
f.close()
result, json_file = rt.update_socrata_contents(json_file, script_name, url, resource)
```
5. `update_socrata_script(script_name, filename, url, resource, script_path)`
This function renames the script, calls the `update_socrata_contents()`, and then writes the new content returned by `update_socrata_contents()`
```python
import retriever as rt
from retriever.lib.defaults import SOCRATA_SCRIPT_WRITE_PATH

script_path = SOCRATA_SCRIPT_WRITE_PATH
resource = rt.find_socrata_dataset_by_id('35s3-nmpm')
filename = resource["id"] + '.csv'
url = 'https://' + resource["domain"] + '/resource/' + filename
script_name = 'socrata-35s3-nmpm'
rt.update_socrata_script(script_name, filename, url, resource, script_path)
```
6. `install_socrata_dataset(engine,
                            name,
                            resource,
                            script_path=SOCRATA_SCRIPT_WRITE_PATH)`
This function installs the socrata dataset into retriever into the engine provided. This function creates the script, then updates it and at last it installs the dataset using that script.

```python
import retriever as rt
from retriever.lib.get_opts import parser
from retriever.engines import choose_engine
from retriever.lib.defaults import SOCRATA_SCRIPT_WRITE_PATH

args = ['install', 'postgres', 'socrata']
args = parser.parse_args(args)
engine = choose_engine(args.__dict__)
engine.use_cache = True
script_path = SOCRATA_SCRIPT_WRITE_PATH
resource = rt.find_socrata_dataset_by_id('35s3-nmpm')
name = 'socrata-35s3-nmpm'
rt.install_socrata_dataset(args, engine, name, resource, script_path)
```

henrykironde commented:
Looks like a style error 🙂, so says flakes. 
we have many of those like,  `if args.dataset is not None:`  >>> `if args.dataset`

Aakash3101 commented:
> Looks like a style error , so says flakes.
> we have many of those like, `if args.dataset is not None:` >>> `if args.dataset`

Sorry my bad @henrykironde

Aakash3101 commented:
@henrykironde I just realised that the `rt.download()` won't download raw data files for those socrata scripts whose script is not yet generated.

henrykironde commented on test/test_retriever.py:
```diff
@@ -243,6 +345,43 @@ def test_dataset_names_upstream():
+def test_socrata_dataset_info():
+    """Check if socrata dataset info returns metadata for a dataset name"""
+    resource = rt.socrata_dataset_info("Building Permits")
+    assert isinstance(resource, list) and (len(resource) != 0) and (resource[0]["id"] != None)
```
`all ([isinstance(resource, list), len(resource), resource[0]["id"]])`
we can also simplify these as above

henrykironde commented on retriever/__main__.py:
```diff
@@ -136,6 +137,42 @@ def main():
+                name_list = socrata_autocomplete_search(args.s)
+                print("Autocomplete suggestions : Total {} results\n".format(
+                    len(name_list)))
+                if len(name_list) > 0:
```
`if len(name_list) > 0:`   to `if len(name_list)`

henrykironde commented on retriever/lib/defaults.py:
```diff
@@ -80,3 +83,11 @@
   doi={10.1371/journal.pone.0065848}
 }
 """
+try:
```
We should try to keep import statements at the top unless there is a good reason to keep them here

henrykironde commented on retriever/__main__.py:
```diff
@@ -136,6 +137,42 @@ def main():
+            elif isinstance(args.s, list) and (not (args.l or args.k or args.v)):
+                try:
+                    import inquirer
+                    from retriever.lib.defaults import INQUIRER_THEME
```
Try to keep the import statements at the top,

henrykironde commented on retriever/lib/engine.py:
```diff
@@ -489,25 +491,72 @@ def download_file(self, url, filename):
                         'user-agent':
-                            'Weecology/Data-Retriever \
-                                            Package Manager: http://www.data-retriever.org/'
+                            'Weecology/Data-Retriever Package Manager: http://www.data-retriever.org/'
```
This block is just the same as the one in download_file  line  L534, find away of refactoring them.

henrykironde commented on retriever/lib/engine.py:
```diff
@@ -489,25 +491,72 @@ def download_file(self, url, filename):
+            print("The download url {} is incorrect!!".format(url))
+            return False
+
+    def download_file(self, url, filename):
```
Let's move `def download_file` above `def download_from_socrata`. We try to cluster functionality together but also sort where possible

henrykironde commented on retriever/lib/install.py:
```diff
@@ -53,6 +54,30 @@ def _install(args, use_cache, debug):
                 print(e)
                 if debug:
                     raise
+    elif args['dataset'].startswith('socrata') and (data_sets_scripts is None):
```
`and data_sets_scripts is None` 
` and not data_sets_scripts`

henrykironde commented on retriever/lib/install.py:
```diff
@@ -53,6 +54,30 @@ def _install(args, use_cache, debug):
+            else:
+                print("{} because it is of type {} and not tabular".format(
+                    resource["error"], resource["datatype"][1]))
+        elif len(resource.keys()) == 0:
```
This should looks to be hanging, may be it should go up. We also do not need to  use == 0, 
>> elif not resourse.keys()

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+    query = " ".join(dataset)
+    url = SOCRATA_BASE_URL + '/autocomplete'
+    params = {'q': query}
+    try:
```
The same Try block as in the other comments, we need to refactor this part

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+
+    except HTTPError as e:
+        print("HTTPError :", e)
+        return []
```
Use return `None` or  `False`. In this case if we return None on the list variable, we can use `if name` to check if its empty.

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+            return []
+        else:
+            datasets = response.json()
+            names = []
```
Since we are returning `names`, let us initialize it at the top of the function so that we do not musk it. It looks hidden

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+        else:
+            data = response.json()
+            result = data["results"]
+            resources = []
```
Most of the issues here are similar to the ones mentioned above. `Try block`, `return []`, hidden `resources` variable

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+            resource = {}
+            data = response.json()
+
+            if len(data["results"]) != 0:
```
if data["results"]:

henrykironde commented on retriever/lib/socrata.py:
```diff
@@ -0,0 +1,260 @@
+        json_file.pop("archived")
+    if ("resources" in json_file.keys()) and (len(json_file["resources"]) != 0) and (
+            "path" in json_file["resources"][0].keys()):
+        json_file["resources"][0].pop("path")
```
if all(["resources" in json_file.keys()), json_file["resources"], "path" in json_file["resources"][0].keys()])

Aakash3101 commented:
Refactoring is complete, do let me know if you want to make any other changes @henrykironde

henrykironde commented:
Why do we get these results ?
 ✗ retriever ls -socrata     
Autocomplete suggestions : Total 0 results

(testenv) ➜  retriever git:(test) ✗

Aakash3101 commented:
> Why do we get these results ?
> 
> ```python
>  ✗ retriever ls -socrata     
> Autocomplete suggestions : Total 0 results
> 
> (testenv) ➜  retriever git:(test) ✗ 
> ```

That's because an empty string is passed to it. And the `socrata_autocomplete_search` function returns an empty list for it.

henrykironde commented:
@Aakash3101 lets us test this together. Whenever you ready  ping me here and on gitter.

Aakash3101 commented:
> @Aakash3101 lets us test this together. Whenever you ready ping me here and on gitter.

Yeah let's test it now

henrykironde commented:
Thank you the work done on this issue @Aakash3101
